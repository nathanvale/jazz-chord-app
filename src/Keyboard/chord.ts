/* eslint-disable no-console */
import { Interval, Note, IntervalVariant } from "teoria";

export class Chord {
  private _rootNote: Note;
  private _voicing: IntervalVariant[];
  private _notes: Note[];

  constructor(rootNote: Note, voicing: IntervalVariant[]) {
    this._rootNote = rootNote; //?
    this._notes = voicing.map((interval) => {
      return rootNote.interval(interval);
    });
    this._voicing = voicing;
  }

  rootNote() {
    return this._rootNote;
  }

  rootNoteName() {
    return this._rootNote.name().toUpperCase() + this._rootNote.accidental();
  }

  notes() {
    return this._notes;
  }

  invertUp() {
    const newNote = this._notes[0].interval(Interval.toCoord("P8"));
    this._notes.push(newNote);
    this._notes.shift();
    const interval = this._voicing.shift();
    if (interval) this._voicing.push(interval);
    return this;
  }

  doubleInvertUp() {
    return this.invertUp().invertUp();
  }

  invertDown() {
    const newNote = this._notes[this._notes.length - 1].interval(
      Interval.toCoord("P-8")
    );
    this._notes.unshift(newNote);
    this._notes.pop();
    const interval = this._voicing.pop();
    if (interval) this._voicing.unshift(interval);
    return this;
  }

  doubleInvertDown() {
    return this.invertDown().invertDown();
  }

  voicing() {
    return this._voicing;
  }

  simple() {
    return this._notes.map((n) => {
      const key = n.toString(true);
      const string = key.charAt(0).toUpperCase() + key.slice(1) + n.octave(); //?
      return string;
    });
  }
}

export function chord(
  root: string,
  octave: number,
  voicing: IntervalVariant[]
) {
  let r, o;
  if (typeof root === "string") {
    r = root.match(/^([a-h])(x|#|bb|b?)/i); //?
    if (r && r[0]) {
      o = typeof octave === "number" ? octave.toString(10) : "4"; //?
      return new Chord(Note.fromString(r[0].toLowerCase() + o), voicing);
    }
  }

  throw new Error("Invalid Chord. Couldn't find note name");
}

//const c = chord("Eb", 4, ["M3", "P5", "M7", "M9"]); //?
// console.log(c.notes());
// console.log(c.simple());
// //console.log(c.invertDown().simple());
// console.log(c.voicing());
// console.log(c.rootNoteName());
// console.log(c.rootNote().accidental());
