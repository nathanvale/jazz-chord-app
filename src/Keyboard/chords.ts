import { IntervalVariant } from "teoria";

export type Key =
  | "C"
  | "G"
  | "D"
  | "A"
  | "E"
  | "B"
  | "F#"
  | "Db"
  | "Ab"
  | "Eb"
  | "Bb"
  | "F";

export type Chord = "maj9" | "min9";

export const keys: Key[] = [
  "C",
  "G",
  "D",
  "A",
  "E",
  "B",
  "F#",
  "Db",
  "Ab",
  "Eb",
  "Bb",
  "F",
];

export const chords: Record<Chord, IntervalVariant[]> = {
  maj9: ["M7", "M9", "M3", "P5"],
  min9: ["m3", "P5", "m7", "M9"],
};
