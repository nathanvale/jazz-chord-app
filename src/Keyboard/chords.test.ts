import { chord } from "./chord";
import { Key, Chord, chords } from "./chords";

const createChordMap = (c: Chord) => (note: Key) => ({
  [note]: chord(note, 3, chords[c]).simple().toString(),
});

it("should create a maj9ww chord", () => {
  expect(createChordMap("maj9")("C")).toMatchSnapshot();
});

it("should create a chord interval map", () => {
  expect(chords).toMatchSnapshot();
});
