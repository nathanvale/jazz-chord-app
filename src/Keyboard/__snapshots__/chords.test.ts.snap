// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a chord interval map 1`] = `
Object {
  "maj9": Array [
    "M7",
    "M9",
    "M3",
    "P5",
  ],
  "min9": Array [
    "m3",
    "P5",
    "m7",
    "M9",
  ],
}
`;

exports[`should create a maj9ww chord 1`] = `
Object {
  "C": "B3,D4,E3,G3",
}
`;
